src/Microsoft.IdentityModel.JsonWebTokens/Experimental/JsonWebTokenHandler.ValidateToken.Internal.cs

```diff
@@ -22,77 +23,114 @@ internal override async Task<ValidationResult<ValidatedToken, ValidationError>>
             CancellationToken cancellationToken)
         {
+            StackFrame? stackFrame;
             if (string.IsNullOrEmpty(token))
             {
+                string key = ValidationError.GetStackFrameKey(memberName: TokenHandler.ValidateTokenAsyncStr);
+                if (!ValidationError.TryGetStackFrame(key, out stackFrame))
+                    stackFrame = ValidationError.GetAsyncStackFrame(key, new StackFrame(0, true));
+
                 return ValidationError.NullParameter(
                     nameof(token),
-                    ValidationError.GetCurrentStackFrame());
+                    stackFrame!);
             }
 
             if (validationParameters is null)
             {
+                string key = ValidationError.GetStackFrameKey(memberName: TokenHandler.ValidateTokenAsyncStr);
+                if (!ValidationError.TryGetStackFrame(key, out stackFrame))
+                    stackFrame = ValidationError.GetAsyncStackFrame(key, new StackFrame(0, true));
+
                 return ValidationError.NullParameter(
                     nameof(validationParameters),
-                    ValidationError.GetCurrentStackFrame());
+                    stackFrame!);
             }
 
             if (token.Length > MaximumTokenSizeInBytes)
             {
+                string key = ValidationError.GetStackFrameKey(memberName: TokenHandler.ValidateTokenAsyncStr);
+                if (!ValidationError.TryGetStackFrame(key, out stackFrame))
+                    stackFrame = ValidationError.GetAsyncStackFrame(key, new StackFrame(0, true));
+
                 return new ValidationError(
                     new MessageDetail(
                         TokenLogMessages.IDX10209,
                         LogHelper.MarkAsNonPII(token.Length),
                         LogHelper.MarkAsNonPII(MaximumTokenSizeInBytes)),
-                    ValidationFailureType.SecurityTokenTooLarge,
-                    ValidationError.GetCurrentStackFrame());
+                    ValidationFailureType.TokenExceedsMaximumSize,
+                    stackFrame!);
             }
 
             ValidationResult<SecurityToken, ValidationError> readResult = ReadToken(token, callContext);
-            if (readResult.Succeeded)
+            if (!readResult.Succeeded)
             {
-                ValidationResult<ValidatedToken, ValidationError> validationResult = await ValidateTokenAsync(
-                    readResult.Result!,
-                    validationParameters,
-                    callContext,
-                    cancellationToken)
-                    .ConfigureAwait(false);
+                string key = ValidationError.GetStackFrameKey(memberName: TokenHandler.ValidateTokenAsyncStr);
+                if (!ValidationError.TryGetStackFrame(key, out stackFrame))
+                    stackFrame = ValidationError.GetAsyncStackFrame(key, new StackFrame(0, true));
+
+                return readResult.Error!.AddStackFrame(stackFrame!);
```

Comment:
If !Succeeded, Error is guaranteed to be not null,  yes?

Do we need to add a NotNull annotation ValidationResult similar to what OperationResult had? Or adjust null analysis settings for net FW?

Reply:
@keegan-caruso I will try NotNull annotation first.